Add advanced configuration page for Cluster Autoscaler (#320)
giantswarm/docs: src/content/guides/advanced-cluster-autoscaler-configuration/index.md

+++
title = "Cluster Autoscaler advanced configuration"
description = "Here we describe how you can customize the configuration of the managed Cluster Autoscaler service in your clusters"
date = "2019-05-09"
type = "page"
weight = 40
tags = ["tutorial"]
+++

# Cluster Autoscaler Advanced Configuration

The [Cluster Autoscaler addon](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) running inside your cluster has additional configuration options and features that can be customized.

> You can customize two of these configuration options on a per cluster basis through a ConfigMap inside your clusters. The ConfigMap is named `cluster-autoscaler-user-values` and is located in the `kube-system` namespace.

Reviewer: ```suggestion
You can customize some of these configuration options on a per cluster basis through a ConfigMap inside your clusters. The ConfigMap is named `cluster-autoscaler-user-values` and is located in the `kube-system` namespace.
```

```nohighlight
$ kubectl -n kube-system get cm cluster-autoscaler-user-values
NAME                                   DATA      AGE
cluster-autoscaler-user-values                    0         11m
```

On cluster creation the ConfigMap is empty and below-mentioned defaults will be applied to the final Cluster Autoscaler deployment. To customize any of the configuration options, you just need to add the respective line(s) in the data field of the user ConfigMap.

__Warning:__ Please do not edit any of the other Cluster Autoscaler related resources. Only the user ConfigMap is safe to edit.

## Scale down utilization threshold

The `scale-down-utilization-threshold` defines the proportion between requested resources and capacity, which under the value Cluster Autoscaler will trigger the scaling down action.

```yaml
data:
  scale-down-utilization-threshold: 0.65
```

Our default value is 65%, which means in order to scale down, one of the nodes has to have less utilization (CPU/memory) than this threshold.

# Scan Interval

> Define what interval is taken to review the state for taking a decision to scale up/down. Our default value is 10 seconds.

Reviewer: ```suggestion
Define what interval is used to review the state for taking a decision to scale up/down. Our default value is 10 seconds.
```

```yaml
data:
  scanInterval: "100s"
```

## Skip system pods

> By default the Cluster Autoscaler will never delete nodes which run pods of kube-system namespace (but daemonset or mirror ones). It can be modified by setting the property to `"false"`.

Reviewer: ```suggestion
By default the Cluster Autoscaler will never delete nodes which run pods of the `kube-system` namespace (except daemonset or mirror ones). It can be modified by setting following property to `"false"`.
```

what do you mean by "mirror ones"?

Author: yeah now I see it is confusing. I meant mirror/static pods (k8s api server,...)

Author: not sure how to say it better, in CA docs called it mirror too

> By default the Cluster Autoscaler will never delete nodes which run pods of the `kube-system` namespace (except daemonset or mirror ones). It can be modified by setting following property to `"false"`.

Author: ```suggestion
By default, the Cluster Autoscaler will never delete nodes which run pods of the `kube-system` namespace (except `daemonset` pods). It can be modified by setting following property to `"false"`.
```

```yaml
data:
  skipNodesWithSystemPods: "false"
```

## Skip pods with local storage

> The Cluster Autoscaler configuration by default deletes nodes with pods using local storage (`hostPath` or `emptyDir`). In case you want to disable this action, you need to set the property to `"true"`.

Reviewer: ```suggestion
The Cluster Autoscaler configuration by default deletes nodes with pods using local storage (`hostPath` or `emptyDir`). In case you want to disable this action, you need to set the following property to `"true"`.
```

+++
title = "Cluster Autoscaler advanced configuration"
description = "Here we describe how you can customize the configuration of the managed Cluster Autoscaler service in your clusters"
date = "2019-05-09"
type = "page"
weight = 40
tags = ["tutorial"]
+++

> # Cluster Autoscaler advanced configuration

Reviewer 2: ```suggestion
# Cluster Autoscaler Advanced Configuration
```

We use title case for the top headline

The [Cluster Autoscaler addon](https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler) running inside your cluster has additional configuration options and features that can be customized.

You can customize two of these configuration options on a per cluster basis through a ConfigMap inside your clusters. The ConfigMap is named `cluster-autoscaler-user-values` and is located in the `kube-system` namespace.

> __Note:__ This feature is only available in more recent cluster versions. To check if your cluster version supports customization through the ConfigMap, you can check if the above-mentioned ConfigMap is present.

Reviewer 2: ```suggestion
__Note:__ This feature is only available in more recent cluster versions. To find out if your cluster version supports customization through the ConfigMap, you can check if the above-mentioned ConfigMap is present.
```

less repitition

```nohighlight
$ kubectl -n kube-system get cm cluster-autoscaler-user-values
NAME                                   DATA      AGE
cluster-autoscaler-user-values                    0         11m
```

On cluster creation the ConfigMap is empty and below-mentioned defaults will be applied to the final Cluster Autoscaler deployment. To customize any of the configuration options, you just need to add the respective line(s) in the data field of the user ConfigMap.

__Warning:__ Please do not edit any of the other Cluster Autoscaler related resources. Only the user ConfigMap is safe to edit.

## Scale down utilization threshold

The `scale-down-utilization-threshold` defines the proportion between requested resources and capacity, which under the value Cluster Autoscaler will trigger the scaling down action.

```yaml
data:
  scale-down-utilization-threshold: 0.65
```

Our default value is 65%, which means in order to scale down, one of the nodes has to have less utilization (CPU/memory) than this threshold.

> # Expander

Reviewer 2: ```suggestion
## Expander
```

Also, can we make that headline more descriptive?

Author: It is a term took from their lingo, the explanation is below

> When there are several node groups managed by the Cluster Autoscaler we can influence which one will be selected to scale up the cluster when it is needed. Our default is `least-waste`. 

Reviewer 2: What is a `node group` here?

Author: A node pool. Currently, it is not released yet, but it can handle node groups, and a customer would like to have the configuration ready for when the feature comes

Reviewer 2: If this is about a node pool then I would also call it node pool instead of node group.

But, as node pools aren't there yet, can we remove this and remind ourselves to add the bit of documentation once we are there?

Author: ok, I used node group because it is the exact term used in the project. I remove it for now then

- `random`: Select a random node group to scale up.
- `most-pods`: Select the node group which is the best fit for all pods.
- `least-waste`: Select the node group which leaves the least fraction of CPU and Memory).
- `price`: Select the node group which is the most cost-effective.
- `priority`: Select the node group based on an user-configured priorities assigned to group names.

```yaml
data:
  expander: "random"
```

> # Scan Interval

Reviewer 2: ```suggestion
## Scan interval
```

Define what interval is taken to review the state for taking a decision to scale up/down. Our default value is 10 seconds.

```yaml
data:
  scanInterval: "100s"
```

> # Skip system pods

Reviewer 2: ```suggestion
## Skip system pods
```

> By default the Cluster Autoscaler will never delete nodes which run pods of kube-system namespace (but daemonset or mirror ones). It can be modified by setting the property to `false`.

Reviewer 2: ```suggestion
By default the Cluster Autoscaler will never delete nodes which run pods of kube-system namespace (but daemonset or mirror ones). It can be modified by setting the property to `"false"`.
```

Adding quotation marks, since in YAML `false` is a boolean and `"false"` is a string. Assuming a string is expected here, based on the examples.

```yaml
data:
  skipNodesWithSystemPods: "false"
```

> # Skip pods local storage

Reviewer 2: ```suggestion
## Skip pods with local storage
```

Also, should this be `nodes` instead of `pods`?

> The Cluster Autoscaler configuration by default deletes nodes with pods using local storage (`hostPath` or `emptyDir`)". In case you want to disable this action, you need to set the property to `true`.

Reviewer 2: ```suggestion
The Cluster Autoscaler configuration by default deletes nodes with pods using local storage (`hostPath` or `emptyDir`). In case you want to disable this action, you need to set the property to `true`.
```
